README.md
> ℹ️ Please note that his plugin is not actively maintained anymore.

# homebridge-withings-air-quality

This plugin retrieves carbon dioxide and temperature levels from a Withings Smart Body Analyzer smart scale and exposes them as HomeKit sensors with [Homebridge](https://github.com/nfarina/homebridge).

## Installation

The plugin can be installed using the command line

```bash
$ npm install -g homebridge-withings-air-quality
```

or by searching **homebridge-withings-air-quality** in the Plugins page of a [Homebridge Config UI](https://github.com/oznu/homebridge-config-ui-x) dashboard.

## Configuration

The Withings Smart Body Analyzer cannot be accessed directly, so the plugin will require [my.withings.com](https://my.withings.com) email address and password in order to fetch the air quality data online.

The plugin also requires the MAC address of the smart scale. It can be found where the batteries are located.

Here is a sample config (you can view [config.sample.json](./config.sample.json) for more details):

```js
{
  "accessories": [
    {
      "accessory": "Withings Air Quality",
      "name": "Bathroom scale",
      "email": "my.withings@email.com",
      "password": "myWithingsPassword",
      "mac": "0a:b2:c3:d4:e5:f6",
      "levels": [
        350,  // Excellent
        1000, // Good
        2500, // Fair
        5000  // Inferior
      ]
    }
  ]
}
```

The `levels` parameter is used to control the information displayed on the air quality sensor. The values are measured in ppm (parts per million). Carbon dioxide level higher than the value for "Inferior" will trigger a warning in HomeKit. For more information, [check out this page](https://support.withings.com/hc/en-us/articles/201489797-Smart-Body-Analyzer-WS-50-Frequently-asked-questions-about-air-quality-measurements).


CHANGELOG.md
## [0.1.4](https://github.com/dhayab/homebridge-withings-air-quality/compare/v0.1.3...v0.1.4) (2021-06-22)


### Bug Fixes

* update connection url ([4ae0418](https://github.com/dhayab/homebridge-withings-air-quality/commit/4ae04183821837d2e28ae4f87eb4eca846b11d9c))



## [0.1.3](https://github.com/dhayab/homebridge-withings-air-quality/compare/v0.1.2...v0.1.3) (2020-02-20)


### Bug Fixes

* add required data to withings login request ([d567a44](https://github.com/dhayab/homebridge-withings-air-quality/commit/d567a44064930257b9d3e78c90f7617d7839f970)), closes [#3](https://github.com/dhayab/homebridge-withings-air-quality/issues/3)
* send right battery information on first tick ([4d62965](https://github.com/dhayab/homebridge-withings-air-quality/commit/4d6296591ac09fb61b7f48912e42a4066fa2af1a))



## [0.1.2](https://github.com/dhayab/homebridge-withings-air-quality/compare/v0.1.1...v0.1.2) (2020-01-24)


### Bug Fixes

* reconnect automatically when session expires ([2366065](https://github.com/dhayab/homebridge-withings-air-quality/commit/236606565e0b7cf756b7de7af603a5f1a383a359)), closes [#2](https://github.com/dhayab/homebridge-withings-air-quality/issues/2)



## [0.1.1](https://github.com/dhayab/homebridge-withings-air-quality/compare/v0.1.0...v0.1.1) (2019-10-12)


### Bug Fixes

* use numbers instead of hap constants in air quality levels ([a00c908](https://github.com/dhayab/homebridge-withings-air-quality/commit/a00c908c6f020093366ff7ba8dcdf93d62725d9f))



# [0.1.0](https://github.com/dhayab/homebridge-withings-air-quality/compare/v0.0.3...v0.1.0) (2019-10-12)


### Features

* replace co2 sensor with air quality sensor ([443351d](https://github.com/dhayab/homebridge-withings-air-quality/commit/443351dd8a190bf7ed76e1e77379d2441f821860))



## [0.0.3](https://github.com/dhayab/homebridge-withings-air-quality/compare/v0.0.2...v0.0.3) (2019-10-12)


### Bug Fixes

* **schema:** add missing 'name' property ([5ccb621](https://github.com/dhayab/homebridge-withings-air-quality/commit/5ccb62122e92b684827c0a7a03d3168603692c29))


### Performance Improvements

* rebuild api and plugin architecture ([c57c2bb](https://github.com/dhayab/homebridge-withings-air-quality/commit/c57c2bbd73c9fd63441db730eabc1330700e46cd))



## 0.0.2 (2019-10-10)





src/lib/api.ts
import superagent from 'superagent';

import {
	ApiResponse, DataType, Device, DeviceApiResponse, EventType, MeasureApiResponse,
} from './api.types';

enum Api {
	Connect = 'https://account.withings.com/connectionwou/account_login',
	Devices = 'https://scalews.withings.com/cgi-bin/association',
	Measure = 'https://scalews.withings.com/cgi-bin/v2/measure',
}

export enum Measure {
	CarbonDioxide = 35,
	Temperature = 12,
}

const agent = superagent.agent();

const INVALID_PARAMS_ERROR = 'Invalid Params';
const TICK_INTERVAL = 15 * 60 * 1000; // tslint:disable-line:no-magic-numbers

export class WithingsApi {
	private device: Device;
	private battery: number;
	private carbondioxide: MeasureApiResponse['body']['series'][0]['data'][0];
	private temperature: MeasureApiResponse['body']['series'][0]['data'][0];

	private subscribers: Array<{ type: EventType, callback: (data: any) => void }> = [];
	private timer: NodeJS.Timeout;

	constructor(
		private readonly email: string,
		private readonly password: string,
		private readonly mac: string,
	) { }

	async init() {
		if (this.timer) {
			return;
		}

		this.timer = setInterval(() => this.tick(), TICK_INTERVAL);
		await this.tick();
	}

	getBatteryLevel() {
		return this.battery;
	}

	getDeviceInfo() {
		return this.device;
	}

	getCarbonDioxide() {
		return this.carbondioxide.value;
	}

	getTemperature() {
		return this.temperature.value;
	}

	on(type: DataType, callback: (data: number) => void): void;
	on(type: 'error', callback: (data: { message: string, error: Error }) => void): void;
	on(type: EventType, callback: (data: any) => void) {
		this.subscribers.push({ type, callback });
	}

	private async connect() {
		const query = await agent.post(Api.Connect)
			.field('email', this.email).field('password', this.password)
			.field('r', 'https://healthmate.withings.com/').field('is_admin', 'f')
		;
		const error = (query.text.replace(/[\n|\r|\t]/g, '').match(/<div class="alert alert-danger"><li>(.+?)<\/li><\/div>/) || [])[1];

		if (error) {
			throw new Error(`Connection failed (${error})`);
		}
	}

	private emit<T>(type: EventType, value: T) {
		this.subscribers
			.filter((subscriber) => subscriber.type === type)
			.forEach(({ callback }) => callback(value))
			;
	}

	private emitError(message: string, error: any) {
		this.emit<{ message: string, error: any }>('error', { message, error });
	}

	private async fetchDeviceInfo() {
		try {
			const query = await agent
				.post(Api.Devices)
				.field('action', 'getbyaccountid').field('enrich', 't')
			;

			const data = JSON.parse(query.text) as ApiResponse;
			if (data.status === 0) {
				this.device = (data as DeviceApiResponse).body.associations
					.find(({ deviceproperties }) => deviceproperties.macaddress.toUpperCase() === this.mac)
				;

				if (this.device) {
					return this.device;
				} else {
					throw new Error(`Could not find device with MAC ${this.mac}`);
				}
			} else {
				throw data.error;
			}
		} catch (e) {
			throw e;
		}
	}

	private async fetchMeasure(type: Measure) {
		try {
			const query = await agent
				.post(Api.Measure)
				.field('action', 'getmeashf').field('meastype', type).field('deviceid', this.device.deviceid)
			;

			const data = JSON.parse(query.text) as ApiResponse;
			if (data.status === 0) {
				const measure = (data as MeasureApiResponse).body.series[0].data[0];
				return measure;
			} else {
				throw data.error;
			}
		} catch (e) {
			throw e;
		}
	}

	private async reset() {
		this.device = null;
	}

	// tslint:disable-next-line:cyclomatic-complexity
	private async tick() {
		if (!this.device) {
			try {
				await this.connect();
			} catch (error) {
				this.emitError('Could not connect', error);
				return;
			}
		}

		try {
			const device = await this.fetchDeviceInfo();
			this.device = device;

			const battery = device.deviceproperties.batterylvl;
			(!this.battery || this.battery !== battery) && this.emit('battery', battery);
			this.battery = battery;

			const carbondioxide = await this.fetchMeasure(Measure.CarbonDioxide);
			(!this.carbondioxide || this.carbondioxide.date !== carbondioxide.date) && this.emit('carbondioxide', carbondioxide.value);
			this.carbondioxide = carbondioxide;

			const temperature = await this.fetchMeasure(Measure.Temperature);
			(!this.temperature || this.temperature.date !== temperature.date) && this.emit('temperature', temperature.value);
			this.temperature = temperature;
		} catch (error) {
			if (error === INVALID_PARAMS_ERROR) {
				this.emitError('Session expired. Will reconnect on next tick.', error);
				this.reset();
			} else {
				this.emitError('Could not fetch data from API', error);
			}
		}
	}
}


src/index.ts
import * as Hap from 'hap-nodejs';

import pkg from '../package.json';
import { WithingsApi } from './lib/api';
import { DataType } from './lib/api.types';

type AirQualityLevel = { level: number, threshold: number, label: string };
type Config = { name: string, email: string, password: string, mac: string, levels?: number[] };
type Homebridge = {
	hap: typeof Hap;
	registerAccessory(pluginName: string, accessoryName: string, constructor: any, configurationRequestHandler?: any): void;
};

let Service: typeof Hap.Service;
let Characteristic: typeof Hap.Characteristic;

const DEFAULT_BATTERY_LOW_LEVEL = 10;
const DEFAULT_AIR_QUALITY_LEVELS: AirQualityLevel[] = [
	{ level: 1, threshold: 350, label: 'Excellent' },
	{ level: 2, threshold: 1000, label: 'Good' },
	{ level: 3, threshold: 2500, label: 'Fair' },
	{ level: 4, threshold: 5000, label: 'Inferior' },
	{ level: 5, threshold: Number.MAX_SAFE_INTEGER, label: 'Poor' },
];

export default function (homebridge: Homebridge) {
	Service = homebridge.hap.Service;
	Characteristic = homebridge.hap.Characteristic;

	homebridge.registerAccessory(pkg.name, 'Withings Air Quality', WithingsScale);
}

class WithingsScale {
	private readonly api: WithingsApi;
	private readonly name = this.config.name;
	private readonly levels: AirQualityLevel[];

	readonly informationService = new Service.AccessoryInformation('', '');
	readonly batteryService = new Service.BatteryService('', '');
	readonly airQualityService = new Service.AirQualitySensor('', '');
	readonly temperatureService = new Service.TemperatureSensor('', '');

	constructor(
		private readonly log: any,
		private readonly config: Config,
	) {
		this.api = new WithingsApi(config.email, config.password, this.formattedMacAddress);
		this.api.on('error', ({ message, error }) => this.log.warn(message, '/', error.message || error));

		if (!config.levels || config.levels.length !== DEFAULT_AIR_QUALITY_LEVELS.length - 1) {
			this.levels = DEFAULT_AIR_QUALITY_LEVELS;
		} else {
			this.levels = DEFAULT_AIR_QUALITY_LEVELS.map((level, index) => {
				return { ...level, threshold: config.levels[index] || level.threshold };
			});
		}

		this.init();
	}

	async init() {
		await this.api.init();

		this.batteryService.setCharacteristic(Characteristic.ChargingState, Characteristic.ChargingState.NOT_CHARGEABLE);
		this.initServiceEvents(
			'battery',
			this.batteryService.getCharacteristic(Characteristic.BatteryLevel),
			() => this.updateBatteryLevel(),
		);

		this.airQualityService.setCharacteristic(Characteristic.Name, 'Air Quality');
		this.initServiceEvents(
			'carbondioxide',
			this.airQualityService.getCharacteristic(Characteristic.AirQuality),
			() => this.updateAirQuality(),
		);

		this.temperatureService.setCharacteristic(Characteristic.Name, 'Temperature');
		this.initServiceEvents(
			'temperature',
			this.temperatureService.getCharacteristic(Characteristic.CurrentTemperature),
			() => this.updateTemperature(),
		);
	}

	getServices() {
		this.informationService
			.setCharacteristic(Characteristic.Name, this.name)
			.setCharacteristic(Characteristic.Manufacturer, 'Withings')
			.setCharacteristic(Characteristic.Model, 'Smart Body Analyzer')
			.setCharacteristic(Characteristic.SerialNumber, this.formattedMacAddress)
			.setCharacteristic(Characteristic.FirmwareRevision, pkg.version)
		;

		return [
			this.informationService,
			this.batteryService,
			this.temperatureService,
			this.airQualityService,
		];
	}

	identify(callback: () => void) {
		this.log('Identify requested!');
		callback();
	}

	private get formattedMacAddress() {
		return this.config.mac.replace(/\W/g, '').match(/\w{2}/g).join(':').toUpperCase();
	}

	private async initServiceEvents(
		type: DataType,
		characteristic: Hap.Characteristic,
		updateFn: () => Hap.CharacteristicValue,
	) {
		this.api.on(type, () => updateFn());
		characteristic.on('get' as Hap.CharacteristicEventTypes, async (cb: Hap.NodeCallback<Hap.CharacteristicValue>) => {
			const value = updateFn();
			cb(null, value);
		});
	}

	private updateBatteryLevel() {
		const value = this.api.getBatteryLevel();
		this.log.debug(`[Battery Level] ${value}%`);
		this.batteryService
			.setCharacteristic(Characteristic.BatteryLevel, value)
			.setCharacteristic(Characteristic.StatusLowBattery, value < DEFAULT_BATTERY_LOW_LEVEL ? 1 : 0)
		;

		return value;
	}

	private updateAirQuality() {
		const value = this.api.getCarbonDioxide();
		const { label, level } = this.levels.find(({ threshold }) => value <= threshold);

		this.log.debug(`[Air Quality] ${label} (${value}ppm)`);
		this.airQualityService
			.setCharacteristic(Characteristic.AirQuality, level)
			.setCharacteristic(Characteristic.CarbonDioxideLevel, value)
		;

		return level;
	}

	private updateTemperature() {
		const value = this.api.getTemperature();
		this.log.debug(`[Temperature] ${value}°C`);
		this.temperatureService
			.setCharacteristic(Characteristic.CurrentTemperature, value)
		;

		return value;
	}
}
